### PR TITLE
Can now drink from chemistry bottles and beakers

### LINKED ISSUE #10614: FEATURE REQUEST: DRINKING FROM BOTTLES
Someone should code drinking from bottles. Also, having bottles break when thrown/over people's head like booze bottles. My potions are too strong for you, traveler. :100:

### PR DESCRIPTION
**Pull request by ghost**
Reapproach of #10631 , now with more integrated suggestions, and less sipping from chempacks

Fixes #10614

:cl:

- rscadd: Can now drink from chem bottles, beakers, etc.
- rscadd: The amount you drink from these bottles/beakers depends on their set transfer rate
- rscadd: If you attempt to force feed somebody from a beaker, the time required scales with the amount you are trying to transfer
- rscadd: To splash a person, or yourself, with the contents of a beaker, you must be in any intent but help

### CODE REVIEW

File: code/modules/reagents/reagent_containers.dm [hunk 1]
```diff
@@ -80,7 +80,64 @@ var/list/LOGGED_SPLASH_REAGENTS = list(FUEL, THERMITE)
...
+
+		if(!iscarbon(user))
+			M.LAssailant = null
+		else
+			M.LAssailant = user
+
+		if(reagents.total_volume)
+			if (ishuman(M))
+				var/mob/living/carbon/human/H = M
+				if(H.species.chem_flags & NO_DRINK)
```
Comment (ComicIronic) [inline code/modules/reagents/reagent_containers.dm]: Don't like the copy paste of this block, put it into its own proc.

Comment (ghost) [reply code/modules/reagents/reagent_containers.dm]: What, like check_drink(H)?

Seems a bit odd for something so specific, but I guess we could have it if anyone adds, say, being able to remove tongues, or jaws

Comment (ComicIronic) [reply code/modules/reagents/reagent_containers.dm]: I mean the entire code for drinking - it's down in imbibe() as well. Anything you write twice in a way that it has to be the same in both cases should be a proc.

**Comment by FudgePucker** [comment]
I don't know about letting players drink 100 units at a time. That seems like you can put in 100000 units of antitoxin to powergame

**Comment by kilozombie** [comment]
IIRC a creature's reagent container can only hold like 300u?

**Comment by Sprok0** [comment]
A human can hold 1000u

**Comment by Allspice** [comment]
Can eat 50u pills anyway, no particular difference on that bit.

File: code/modules/reagents/reagent_containers.dm (hunk 1 above)
Comment (ghost) [reply code/modules/reagents/reagent_containers.dm]: Fair point, and good spot on the weirdness going on there. Sorted now

**Comment by Zth--** [comment]
What about the lids? Did you change it or they're still the same?

**Comment by ghost** [comment]
The same as usual. For lids, you click it in-hand. For drinking, you click on your or anothers sprite

**Comment by Zth--** [comment]
I'm asking because I do everything with the "Z" key. I drink alcohol with the Z key, and put beaker's lid with the Z key. I guess the default will still be the lid for beakers.

Don't get me wrong, it has sense to be this way, I just wasn't sure if you changed it or not

**Comment by ghost** [comment]
Z is attack_self IIRC, so should still work

File: code/modules/reagents/reagent_containers.dm [hunk 2]
```diff
@@ -290,4 +336,32 @@ var/list/LOGGED_SPLASH_REAGENTS = list(FUEL, THERMITE)
...
+
+	return 1
+
+/obj/item/weapon/reagent_containers/proc/can_drink(mob/user)
+	if (ishuman(user))
+		var/mob/living/carbon/human/H = user
+		if(H.species.chem_flags & NO_DRINK)
+			reagents.reaction(get_turf(H), TOUCH)
+			H.visible_message("<span class='warning'>The contents in [src] fall through and splash onto the ground, what a mess!</span>")
+			reagents.remove_any(amount_per_transfer_from_this)
```
Comment (Iamgoofball) [inline code/modules/reagents/reagent_containers.dm:364]: make it react with the ground below it first, if I chug thermite as a skeleton and it falls to the ground I should be able to light it

Comment (ghost) [reply code/modules/reagents/reagent_containers.dm:364]: Isnt that what the above reagents.reaction(get turf(h), touch)

Comment (ghost) [reply code/modules/reagents/reagent_containers.dm:364]: Does?

Comment (Intigracy) [reply code/modules/reagents/reagent_containers.dm:364]: Yes, ignore goofball.

Comment (ghost) [reply code/modules/reagents/reagent_containers.dm:364]: goof ya goof

**Comment by kol1th** [comment]
I love you 👍

File: code/modules/reagents/reagent_containers.dm [hunk 3]
```diff
@@ -80,7 +80,53 @@ var/list/LOGGED_SPLASH_REAGENTS = list(FUEL, THERMITE)
...
+	else if(ishuman(M))
+		user.visible_message("<span class='danger'>[user] attempts to feed [M] \the [src].</span>", "<span class='danger'>You attempt to feed [M] \the [src].</span>")
+
+		var/drink_delay = 30*(amount_per_transfer_from_this/5) //So normal = 1
+		if(!do_mob(user, M, drink_delay))
+			return 1
+
+		user.visible_message("<span class='danger'>[user] feeds [M] \the [src].</span>", "<span class='danger'>You feed [M] \the [src].</span>")
+
+		M.attack_log += text("\[[time_stamp()]\] <font color='orange'>Has been fed [src.name] by [user.name] ([user.ckey]) Reagents: [reagentlist(src)]</font>")
```
Comment (ComicIronic) [inline code/modules/reagents/reagent_containers.dm]: Replace from this down to setting LAssailant with a call to add_attacklogs, it handles all the stuff for you.

**Review by ComicIronic — CHANGES_REQUESTED** [review]

File: code/modules/reagents/reagent_containers.dm [hunk 4]
```diff
@@ -80,7 +80,54 @@ var/list/LOGGED_SPLASH_REAGENTS = list(FUEL, THERMITE)
...
+	else if(ishuman(M))
+		user.visible_message("<span class='danger'>[user] attempts to feed [M] \the [src].</span>", "<span class='danger'>You attempt to feed [M] \the [src].</span>")
+
+		var/drink_delay = 30*(amount_per_transfer_from_this/5) //So normal = 1
+		if(!do_mob(user, M, drink_delay))
+			return 1
+
+		user.visible_message("<span class='danger'>[user] feeds [M] \the [src].</span>", "<span class='danger'>You feed [M] \the [src].</span>")
+
+		add_attacklogs(user, M, "force-fed", src, "amount:[amount_per_transfer_from_this], container containing [reagentlist(src)]")
```
Comment (ghost) [inline code/modules/reagents/reagent_containers.dm:116]: Done, but it might get a bit obnoxious

Comment (ComicIronic) [reply code/modules/reagents/reagent_containers.dm:116]: No, I meant remove the LAssailant stuff as well - add_attacklogs does that too.

Comment (ghost) [reply code/modules/reagents/reagent_containers.dm:116]: ???

File: code/modules/reagents/reagent_containers.dm [hunk 5]
```diff
@@ -80,7 +80,54 @@ var/list/LOGGED_SPLASH_REAGENTS = list(FUEL, THERMITE)
...
+		/*M.attack_log += text("\[[time_stamp()]\] <font color='orange'>Has been fed [src.name] by [user.name] ([user.ckey]) Reagents: [reagentlist(src)]</font>")
+		user.attack_log += text("\[[time_stamp()]\] <font color='red'>Fed [M.name] by [M.name] ([M.ckey]) Reagents: [reagentlist(src)]</font>")
+		log_attack("<font color='red'>[user.name] ([user.ckey]) fed [M.name] ([M.ckey]) with [src.name] (INTENT: [uppertext(user.a_intent)])</font>")*/
+
+		if(!iscarbon(user))
+			M.LAssailant = null
+		else
+			M.LAssailant = user
```
Comment (ghost) [inline code/modules/reagents/reagent_containers.dm:124]: So, this

Comment (ComicIronic) [reply code/modules/reagents/reagent_containers.dm:124]: Yes.

**Comment by ghost** [comment]
fork broke, head to #12719

**Pull request closed without merging**